File: packages/program-design/Leetcode/hot100/数据结构/二叉树/二叉树的右视图.ts
import type { TreeNode } from "./types";

// 可以用层次遍历 + 取每一层最右边
function rightSideView(root: TreeNode | null): number[] {
  if (root === null) {
    return []
  }

  const res: number[] = [];
  const q: TreeNode[] = [root];
  while (q.length) {
    const len = q.length;

    let lastVal = 0;
    for (let i = 0; i < len; i++) {
      const top = q.shift()!;

      lastVal = top.val;

      if (top.left) q.push(top.left);
      if (top.right) q.push(top.right);
    }

    res.push(lastVal);
  }

  return res;
};

File: packages/program-design/alog/first/数据结构/二叉树/py/二叉树的右视图.py
# Definition for a binary tree node.
import queue
from typing import Optional, List


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        if not root:
            return []

        ans = []
        q = queue.Queue()
        q.put(root)

        while not q.empty():
            size = q.qsize()
            for i in range(size):
                node = q.get()
                if node.left:
                    q.put(node.left)
                if node.right:
                    q.put(node.right)
                if i == size - 1:
                    ans.append(node.val)

        return ans
